docs/docs/server-middleware.md

```diff
@@ -0,0 +1,645 @@
+# Server Middleware
+
+Http4s includes some middleware out of the box in the `org.http4s.server.middleware`
+package. Some of it is documented in its own page:
+
+* [Authentication]
+* Cross Origin Resource Sharing ([CORS])
+* Response Compression ([GZip])
+* [HSTS]
+* [CSRF]
+
+We'll describe and provide examples for the remaining middleware, but first we set up our service:
+```scala mdoc:silent
+import cats.effect._
+import cats.syntax.all._
+import org.typelevel.ci._
+import org.http4s._
+import org.http4s.dsl.io._
+import org.http4s.implicits._
+import cats.effect.unsafe.IORuntime
+import scala.concurrent.duration._
+import cats.effect.std.Random
+import fs2.Stream
+
+implicit val runtime: IORuntime = cats.effect.unsafe.IORuntime.global
+
+object NameQueryParamMatcher extends QueryParamDecoderMatcher[String]("name")
+
+val service = HttpRoutes.of[IO] {
+  case GET -> Root / "bad" => BadRequest()
+  case GET -> Root / "ok" => Ok()
+  case r @ POST -> Root / "post" => r.as[Unit] >> Ok()
+  case GET -> Root / "b" / "c" => Ok()
+  case POST -> Root / "queryForm" :? NameQueryParamMatcher(name) => Ok(s"hello $name")
+  case GET -> Root / "wait" => IO.sleep(10.millis) >> Ok()
+  case GET -> Root / "boom" => IO.raiseError(new RuntimeException("boom!"))
+  case r @ GET -> Root / "reverse" => r.as[String].flatMap(s => Ok(s.reverse))
+  case GET -> Root / "forever" => IO(
+    Response[IO](headers = Headers("hello" -> "hi"))
+      .withEntity(Stream.constant("a").covary[IO])
+  )
+  case r @ GET -> Root / "doubleRead" => (r.as[String], r.as[String])
+    .flatMapN((a, b) => Ok(s"$a == $b"))
+  case GET -> Root / "random" => Random.scalaUtilRandom[IO]
+    .flatMap(_.nextInt)
+    .flatMap(random => Ok(random.toString)) 
+}
+
+val okRequest = Request[IO](Method.GET, uri"/ok")
+val badRequest = Request[IO](Method.GET, uri"/bad")
+val postRequest = Request[IO](Method.POST, uri"/post")
+val waitRequest = Request[IO](Method.GET, uri"/wait")
+val boomRequest = Request[IO](Method.GET, uri"/boom")
+val reverseRequest = Request[IO](Method.GET, uri"/reverse")
+```
+The provided examples will use the pattern `service(request)` to show the effects of the middleware, and sometimes
+drain the response body with `response.as[Unit]`, this is necessary when the middleware depends on the request's
+completion (like `MaxActiveRequests`) but in other cases we might choose to omit that. When testing middleware
+in this manner (without running a client and a server) be aware of how it interacts with the response body and
+consume it if necessary. 
+
+Also note that these examples might use non-idiomatic constructs like `unsafeRunSync` and mutable collections for
+conciseness.
+
```

Comment:
Maybe we can insert a navigation tree here as a sort of table-of-contents.
```
@:navigationTree {
  entries = [ { target = "#" } ]
}
```

https://planet42.github.io/Laika/0.18/03-preparing-content/02-navigation.html#the-navigationtree-directive

Reply:
(Or higher up to, wherever you think :)

Reply:
Good call, I'm in favour of that 👍🏻

Reply:
Done!